### ISSUE TITLE
Reference counting of DOM nodes

### ISSUE THREAD

**Issue by phoerious** [lexbor/lexbor#132]
Sorry to abuse the issue tracker as a support platform, but is there a way to reference-count individual DOM nodes or at least check if a given `lxb_dom_node_t*` pointer is still valid or has been destroyed in the meantime? If not, would this be a feature that could be added? The only thing I can find is `lexbor_mraw_reference_count()`, but that only seems to work for the whole document memory block.

I'm writing a Python wrapper in which `DOMNode` instances hold references to DOM nodes and since the lifetime of this instance is independent of the actual `lxb_dom_node_t*` pointer, I have no way of checking whether the element is still valid and all I can do is to warn the user not to use existing `DOMNode`instances after DOM manipulations. I checked whether Selectolax has any smart logic to prevent dangling pointers, but it looks like Selectolax is not taking any precautions either.

**Comment by lexborisov** [comment]
@phoerious 

> Sorry to abuse the issue tracker as a support platform

This is fine. Where else can you ask for additional features?! :)

> reference-count individual DOM nodes or at least check

I planned to do this after adding CSS Styles.

> I'm writing a Python wrapper in which DOMNode instances hold references to DOM nodes and since the lifetime of this instance is independent of the actual lxb_dom_node_t* pointer, I have no way of checking whether the element is still valid and all I can do is to warn the user not to use existing DOMNodeinstances after DOM manipulations.

It seems that you will not be able to do this even with links. If you destroy a `lxb_dom_node_t *`, you still won't get the number of links from it.

I already forgot how to write modules for C in Python.
It seemed to me that Python itself monitors its objects and it has a callback for destruction (when the user does not need it, when all links to it are finished. Сallback will be called).

That is, you must return one object for the same lexbor DON Node until the user destroys it.
For example:
```Python
head_1 = document.head
head_2 = document.find_by_name("head");

head_1 == head_2 # true
```
In Python C some like:
```C
py_node *
py_document_head(..., lxb_html_document_t *document)
{
    lxb_dom_node_t *node;
    lxb_html_head_element_t *head;

    head = lxb_html_document_head_element(document);
    if (head == NULL) {
        return NULL;
    }

    node = lxb_dom_interface_node(head);

    /* We always check for the presence of a Python node. */

    if (node->user == NULL) {
        node->user = py_node_alloc(..., node);
        if (node->user == NULL) {
            /* Oh my God the Python out of memory. */
        }
    }

    return node->user;
}

py_node *
py_callback_to_destroy_node(..., py_node)
{
    /* If the user asks for an object again, see the code above. */
    py_node->ref_to_lxb_node->user = NULL;
}
```

**Comment by phoerious** [comment]
Okay, I didn't know you could use the `user` pointer for your own purposes, but it doesn't really solve my problem either. My logic is actually inverse to yours: I am constructing `DOMNode` objects, which simply hold a pointer to the Lexbor DOM node. I can reconstruct that object at any point, as the only data that it holds is that pointer. Once the `DOMNode` object goes out of scope, I check whether the referenced DOM node is orphaned and if yes, I delete it. If not, I don't do anything, since it will be deleted automatically when the DOM tree itself goes out of scope.

The two problems that can occur here are:

- User still holds a reference to a `DOMNode` instance, whose `lxb_dom_node_t` has been destroyed somewhere else (e.g. because the parent tree got deleted)
- User copies the `DOMNode` object and deletes one of them

I could solve the second issue by ensuring that `DOMNode` is a singleton for each individual `lxb_dom_node_t*` pointer, but I cannot solve the first issue without maintaining an expensive mapping of all pointers to all DOM nodes and checking the whole subtree each time I manipulate a node.

**Comment by lexborisov** [comment]
@phoerious 

You are trying to implement a very expensive and slow binding/wrapper approach.
Above I described the fastest approach. Everything is created only when needed. At the same time, you have complete control over memory and all elements.

**Comment by phoerious** [comment]
Oh, I am not creating any objects that are not needed. `DOMNode` is only the public access point for the user. Everything else is done internally by Lexbor.

**Comment by phoerious** [comment]
BTW:
```
HTML parser benchmark <title> extraction:
=========================================
Resiliparse (Lexbor):  42015 documents in 36.55s (1149.56 documents/s)
Selectolax (Lexbor):   42015 documents in 37.46s (1121.52 documents/s)
Selectolax (MyHTML):   42015 documents in 53.82s (780.72 documents/s)
BeautifulSoup4 (lxml): 42015 documents in 874.40s (48.05 documents/s)
```
:heart:

**Comment by lexborisov** [comment]
@phoerious 

Awesome!
Resiliparse is public wrapper?

**Comment by phoerious** [comment]
Yes. It's a web archive processing library and the wrapper is one new module.

Here's the repository: https://github.com/chatnoir-eu/chatnoir-resiliparse
and here are the docs (stable version, no wrapper docs yet): https://resiliparse.chatnoir.eu/en/stable/

The new module hasn't been merged yet and I am still writing the documentation, but I'll tag a new version as soon as that is done and #125 has been released (so the CI can build the binaries and API docs). In the meantime, here is the module code:

https://github.com/chatnoir-eu/chatnoir-resiliparse/blob/feature/html-parsing/resiliparse/parse/html.pyx

In the end, I decided to use the `user` pointer for ensuring that the wrapper objects are indeed singletons (although that's not necessarily important), but the primary issue remains: I cannot track whether the underlying DOM node is still alive.

**Comment by rushter** [comment]
I think some kind of smart ID could fix this problem.

We need to have a numerical ID for each node, but such an ID must have one important property: it should be possible to check if a given ID is not inserted into any parent node with another ID.

So, if a user deletes a big subtree, we keep the root ID of the subtree. Next time, if a user tries to access any child node that has already been wrapped, we need to check that the ID of the child node does not depend on the removed parent.

Not sure if such IDs exist for the tree data structures.

**Comment by phoerious** [comment]
I would avoid anything that requires tree traversal (up or down). All I am interested in is whether the pointer that I currently hold is still valid. That could be done easily with some sort of weak smart pointer concept, where the user doesn't take ownership, but can rely on the reference either pointing to a valid object or NULL. Ideally that smart pointer type can be cast to a raw pointer, so you can use it as a drop-in replacement (similar to how Qt's QPointer works).

**Comment by rushter** [comment]
> I would avoid anything that requires tree traversal (up or down)

This is why I called this smart ID, it should be possible to check if a child node is inside any given parent without making any traversals just given two IDs (any parent/root, child)

**Comment by phoerious** [comment]
I don't think you really need something fancy if you hold a weak smart pointer reference to a node in your wrapper. If the pointer is != NULL, you know that your node still exists, otherwise your wrapper is stale and you should throw an exception or return None or something.

The same applies to the parent pointer, but here you don't really need a smart pointer if Lexbor takes care of resetting all parent pointers to NULL on the removal of a node (in the end, the effect is the same, the pointer is either valid or NULL). Then if your wrapper goes out of scope and the referenced node does not have a parent, you deallocate the subtree (that is possible already and Resiliparse does exactly that). What's missing, though, is that a wrapper referencing another node in this now-deallocated subtree would need to update its references to NULL. That would happen automatically if the pointer is a weak smart reference.

You could argue that you don't want to delete the whole subtree if there is still a wrapper holding a reference to a (grand-)child node, in which case it's a bit more complicated, because you would need a reverse check which nodes are being held by a wrapper. All nodes without a wrapper would need to be deallocated and all wrapped nodes and their subtree would be retained but orphaned. I don't think that's possible without holding a strong (owning) smart pointer in the wrapper and instructing your recursive delete to deallocate only nodes where the reference count is exactly 1. So for simplicity and speed, I would prefer to simply deallocate the whole subtree if you delete a node. If you want to retain a child, you can easily remove it from the DOM first before deleting the parent.

For that reason, Resiliparse has no non-recursive `decompose()` method (unlike Selectolax), because non-recursive node deletion will necessarily lead to memory leaks where orphaned subtrees without wrappers remain in memory until the owning document is deallocated.

**Comment by phoerious** [comment]
I just noticed that I also cannot destroy unparented nodes in a `DOMNode` wrapper object due to this bug/missing feature.
Let's take this example:

```python
tree = HTMLTree.parse("...")
new_element = tree.create_element('p')
new_text = tree.create_text_node('Hello World!')
new_element.append_child(new_text)
# SEGFAULT
```

This will crash if I do a recursive delete in the destructor of the DOM node wrapper object, because `new_element` will delete the node that is held by `new_text`, which will then have a dangling pointer. If I reverse the order of initialization, it works, because `new_text` will be deleted before `new_element`.
```python
tree = HTMLTree.parse("...")
new_text = tree.create_text_node('Hello World!')
new_element = tree.create_element('p')
new_element.append_child(new_text)
# OK
```

There is no solution to this without reference counting. The only thing I can do is not to delete orphaned nodes, which will leak memory until the tree is destroyed (which is probably an edge case, but ugly nonetheless). The two potential solutions that might come to mind don't work for these reasons:

1. **Don't delete recursively:** 
  Assumes that every element in the subtree has a wrapper object, which is not necessarily the case. Unwrapped objects would stay in memory without a recursive delete.
2. **Implement a custom recursive delete that resets all wrappers of child nodes:**
  Only works in this explicit case, but not if the subtree has been modified by other means of which the wrapper isn't aware (e.g. by setting innerHTML of some node).

I hope that emphasizes the importance of a reference counting feature for DOM nodes.

**Comment by phoerious** [comment]
@lexborisov Any progress on this?

**Comment by lexborisov** [comment]
@phoerious 

I'll deal with this soon.

**Comment by phoerious** [comment]
How's the status? I have a pretty complete DOM API in Rust as a wrapper around lexbor now, but this thing is still bothering me.

See: https://github.com/chatnoir-eu/chatnoir-resiliparse/tree/rust/resiliparse/resiliparse_rust/resiliparse-common/src/parse/html

**Comment by phoerious** [comment]
I've checked the source code to find the best way reference counting could be implemented and I saw that some objects in the CSS parser module already have a ref_count attribute. Unfortunately, none of the DOM interfaces has this. Adding one to `lxb_dom_node` would definitely break ABI compatibility and the counter would need to be updated in every call to at least `lxb_dom_document_create_interface` and `lxb_dom_document_destroy_interface`.

I think this is doable, but it doesn't solve the problem of reallocs where existing pointers become dangling even if the object itself still exists. Fortunately, this doesn't seem to happen anywhere for any of the raw memory chunks and is used only for strings.

Another approach would be to add a `lexbor_hash_t` to `lxb_dom_document` which keeps track of created and destroyed objects within that document.

I can try to make a PR for this, but would like to know your favourite approach.

**Comment by lexborisov** [comment]
Hi @phoerious 

Yes, I understand your need for it, but at the current stage I am doing layout and will implement the dependency counter from the needs of it. It took me a long time to figure out how it should be arranged, and it's definitely not via hash. As soon as the first rendering is ready, it will be clear what it should look like. I'm not sure yet.
Any ideas are welcome!

**Comment by phoerious** [comment]
I've made a quick and dirty draft of it here, which I'm testing at the moment. I believe this works for my use case, but it can certainly be improved. Since this changes the size of the `lxb_dom_node` struct, it will crash if you use an incompatible library version.

See: https://github.com/phoerious/lexbor/commit/65a7e5e50780845683de4ffe78bdd693ae211fcb

**Comment by phoerious** [comment]
I spoke too soon. This solution gets me there most of the way, but when I deallocate attributes, I get stuck in an endless loop somewhere.

**Comment by lexborisov** [comment]
@phoerious 

I'm not ready to accept such changes yet, although they seem to be true. 
I'll come back to this topic right after layout.

**Comment by phoerious** [comment]
I'm fine with running this as a custom patch for now, it just makes it more difficult for others to compile from sources.

**Comment by calvin2021y** [comment]
> Fixed it. The new approach requires more code changes, but maintains the reference counts directly inside the `*_interface_create` functions, which is more robust. This approach works for me and ASAN doesn't report any violations or leaks.

This look good.

I think maybe use `ptrdiff_t` replace `size_t` as `ref_count` type.

**Comment by alfsb** [comment]
Hi. I found this PR by chance, and probably arrived too late into the conversation.

The singular problem here, and in all related issues above, is that an "external observer" needs to know if a C pointer was invalidated *or moved*. Reference counting only fixes the issue partially. Also, transversals need to be avoided.

The problem is akin to Lexbor having [MutationRecord](https://dom.spec.whatwg.org/#interface-mutationrecord)s, but for C object pointers. Let's call it a `LexborMemoryMap`, that may be implemented in Lexbor in the end, but can also be implemented in clients to test alternatives. Here I will discuss here one implementation in the external observers.

The `LexborMemoryMap` resource/object has two methods, `acquire()` and `release()`. Each time an external `DOMNode` takes note of a `lxb_dom_node_t*` address, it calls `acquire()` with this address, and receives one opaque address, and only stores this opaque address. And each time a `DOMNode` forgets a lxb node, it calls `release()` on the opaque address.

Let's call the returned opaque address as `nativeIndirect` for now. The important thing is that the `nativeIndirect` object itself resides outside of the observed object, no breaking ABI here.

What the `LexborMemoryMap` does is to manage a collection of `nativeIndirect`s, having fields like these:
```c
struct lexbor_memory_nativeIndirect
{
    uintptr_t address;
    int count;
}
```
Calling `acquire()` with the observed **object** address will:

1. Binary search all `nativeIndirect` in `LexborMemoryMap`, for one that has the same object address;
2. If found, increment count, and return the address of alive.
3. If not found, create a new record, set the address, set count = 1, return the address of `nativeIndirect`.

Calling `release()` with the `nativeIndirect` will:

1. Dereference and decrease count by one;
2. If < 0, hard error.

The logic above is that in C, object address is identity, but also C objects can be destroyed. So, in a long running process addresses will be reused, then identity from address is not universal. Addresses of objects *under the control of Lexbor* can be destroyed, moved, and *created or moved in a reused address*. This is the problem to solve.

So instead of using these "unstable" identity addresses, here we create a separate object that is only destroyed when *both* sides become uninterested. The external, observer by some garbage collector scheme (be it reference counting of mark and sweep) and only updates this side (`count`), and Lexbor C using C style memory management that cares only for his side (`address`).

Code like this assumes that `acquire()` is the less frequent case, so they are not O(1), but now two others things are O(1) time:

1. The external observer checking if an object still alive is just `if ( indirect->address != NULL )`;
~2. The processing of invalidation of referecended objects is just `*stillAlive = 0;`.~

While invalidation can be somewhat less frequent, it can occur *in masse*, for example, when a document with millions of nodes is destroyed, ~so this case also gains priority in having O(1) time behaviour.~ So if there is a way to store a bit flag anywhere in the observed nodes, to indicate that the node is observed and should call the callbacks below, it will make things way faster.

Without this flag, another weaker optimization is possible. One key observation is that Lexbor can be used "alone", that is, without external observers. In this case, the memory callback address will always be NULL. So no performance is lost. When Lexbor is used in conjunction with external node observers, then the memory callback will be filled, and mass node invalidation may cause some performance impact. Binary searching may keep the impact on O(N) level.

---

Note this will not cover the case of a valid object that is still valid after being *moved* in memory. For completeness, the general solution would be the Lexbor having something like a `void memory_callback( uintptr_t oldObj , uintptr_t newObj )` function (or directly updating its internal `LexborMemoryMap`), that external observers can hook, and:

* `newObj == NULL` means object invalidation.
* Otherwise, the object moved.

--

For external observers, the `if ( indirect->address != NULL )` may be implemented in a macro, something like `get_lxb_node_temporary_address()`, to keep things opaque and "typed". This "function" returns a nullable pointer, so the pattern:
```c
tempAddress = get_lxb_tempoary_address( opaque );
if ( tempAddress != NULL )
    doTemporaryThings();
```

---

Edited to simplify the struct to two fields, striked parts in error.
